File: src/android/ansight-core/src/main/kotlin/ai/ansight/runtime/SessionVisualTreeCaptureRegistry.kt
package ai.ansight.runtime

import org.json.JSONObject

object SessionVisualTreeCaptureRegistry {
    private val lock = Any()
    private var provider: (AndroidToolExecutionContext) -> List<JSONObject> = {
        listOf(AndroidUiEvidence.visualTree())
    }

    @JvmStatic
    fun setProvider(provider: ((AndroidToolExecutionContext) -> List<JSONObject>)?) {
        synchronized(lock) {
            this.provider = provider ?: { listOf(AndroidUiEvidence.visualTree()) }
        }
    }

    internal fun capture(context: AndroidToolExecutionContext): List<JSONObject> {
        val current = synchronized(lock) { provider }
        return current(context)
    }
}

File: src/dotnet/Ansight.Core/Screenshot/SessionVisualTreeCaptureRegistry.cs
using System.Text.Json.Nodes;

namespace Ansight.Screenshot;

/// <summary>
/// Connects optional visual-tree providers to screenshot-and-tree session capture.
/// </summary>
public static class SessionVisualTreeCaptureRegistry
{
    private static readonly Lock gate = new();
    private static Func<CancellationToken, Task<IReadOnlyList<JsonObject>>>? provider;

    /// <summary>
    /// Sets the process-wide provider used to capture visual trees alongside screenshots.
    /// </summary>
    public static void SetProvider(Func<CancellationToken, Task<IReadOnlyList<JsonObject>>>? provider)
    {
        lock (gate)
        {
            SessionVisualTreeCaptureRegistry.provider = provider;
        }
    }

    internal static Task<IReadOnlyList<JsonObject>> CaptureAsync(CancellationToken cancellationToken)
    {
        Func<CancellationToken, Task<IReadOnlyList<JsonObject>>>? provider;
        lock (gate)
        {
            provider = SessionVisualTreeCaptureRegistry.provider;
        }

        return provider?.Invoke(cancellationToken)
               ?? Task.FromResult<IReadOnlyList<JsonObject>>(Array.Empty<JsonObject>());
    }
}
